src/compose/service-manager.coffee

```
@@ -0,0 +1,276 @@
+					.then (deviceType) ->
+						throw new Error("Application architecture incompatible with #{deviceType}: exec format error")
+				else
+					# rethrow the same error
+					throw err
+			.catch (err) =>
+				# If starting the container failed, we remove it so that it doesn't litter
+				container.remove(v: true)
+				.finally =>
+					@logger.logSystemEvent(logTypes.startServiceError, { service, error: err })
+					throw err
+			.then =>
+				@logger.attach(@docker, container.id, service.serviceId)
+		.tap =>
+			if alreadyStarted
+				@logger.logSystemEvent(logTypes.startServiceNoop, { service })
+			else
+				@logger.logSystemEvent(logTypes.startServiceSuccess, { service })
+		.then (container) ->
+			service.running = true
+		.finally =>
+			@reportChange(containerId)
+
+	# Gets all existing containers that correspond to apps
+	getAll: (extraLabelFilters = []) =>
+		filters = label: [ 'io.resin.supervised' ].concat(extraLabelFilters)
+		@docker.listContainers({ all: true, filters })
+		.then (containers) =>
+			Promise.mapSeries containers, (container) =>
+				@docker.getContainer(container.Id).inspect()
+				.then(Service.fromContainer)
+
+	# Returns an array with the container(s) matching a service by appId, commit, image and environment
+	get: (service) =>
+		@getAll("io.resin.service_id=#{service.serviceId}")
+		.then (services = []) ->
+			return _.filter services, (currentService) ->
+				currentService.isSameContainer(service)
+
+	getStatus: =>
+		@getAll()
+		.then (services) =>
+			status = _.clone(@volatileState)
+			_.forEach services, (service) ->
+				status[service.containerId] ?= _.pick(service, [ 'appId', 'imageId', 'status', 'releaseId', 'commit' ])
+			return _.values(status)
+
+	getByDockerContainerId: (containerId) =>
+		@docker.getContainer(containerId).inspect()
+		.then (container) ->
+			if !container.Config.Labels['io.resin.supervised']?
+				return null
+			return Service.fromContainer(container)
+		.catchReturn(null)
```

Why is this hiding an error?

---

The idea was that a null result meant the container wasn't found (we've done this in other places in the past), but it's true that is ugly, so I changed it to not hide the error, and moved the catchReturn to the caller.